### DISCUSSION TITLE
Overview of computed fields 🧮

### DISCUSSION THREAD

**Discussion by caniko** [pydantic/pydantic#4648] [Pydantic V2]
There has been a lot of discussion on `computed fields`, the feature has been deferred to pydantic v2 release. I hope to consolidate all the information in one place to make sure none gets lost. Please post your suggestion lost or new here, links to previous discussions are a pluss.

### Motivation
#935, #1241, #2313, #4609

### Suggestion
```python
...
    @computed_field
    @cached_property OR @property
    def hard_compute():
        <very hard problem>
...
```

### Solutions (ordered by attention)
By @PrettyWood #2625
By @jpribyl #3160

### Conclusions
I include links for additional context, click on "By"!

[By](https://github.com/pydantic/pydantic/pull/2625#issuecomment-1152176712) @caniko:
- [ ] Testing with [`frozen=true`](https://github.com/samuelcolvin/pydantic/pull/2625#issuecomment-992359653)
- [ ] [Validation](https://github.com/samuelcolvin/pydantic/pull/2625#issuecomment-1030826268) feature, and testing
- [ ] Testing that combines `classmethod` with `computed_property`:
  - [ ] Combined with `property`
  - [ ] Combined with `cached_property`
- [ ] Method for and testing flushing cache of `cached_property` based `computed_fields` of both object and class

[By](https://github.com/pydantic/pydantic/pull/2625#issuecomment-1150476569) @Bobronium:
- [ ] Fix typing: https://github.com/samuelcolvin/pydantic/pull/2625#pullrequestreview-649988316
- [ ] Fix inheritance: https://github.com/samuelcolvin/pydantic/pull/2625#issuecomment-1093799303
  - [ ] raise `AttributeError` in `ComputedField.__getattr__` when condition is not met
  - [ ] add tests that include inheritance
- [ ] Decide whether to support `@abstractmethod` (IMO, no, see https://github.com/samuelcolvin/pydantic/pull/2625#issuecomment-1150460832)

Newer ideas:
- [ ] Testing with `async`, use `anyio`

#### Include a keyword argument for computing all computed fields when running `model.dict()` or `model.json()`.

Example:
```python
model.dict(with_computed_fields=True)
```
Will evaluate and provide `dict` with all computed fields _computed_.

#### Shorthand notation
Notation for `property` and `cached_property`:
```python
    @cached_computed_field_property
    ...
```
And:
```python
    @computed_field_property
    ...
```
Should be used when `computed_fields` frequent. Let us avoid decorator-hell.

**Comment by samuelcolvin** [comment]
Thanks so much for this.

This will really help when working on this for V2.

In future probably better to use issues for stuff like this, you can choose the blank template to avoid all the fields if it's easier. Because discussions are less visible and don't integrate well with pull requests, I think best if we leave discussions for questions etc.

**Reply by caniko** [reply]
My pleasure! Good point, I made an [issue](https://github.com/pydantic/pydantic/issues/4650), keeping it closed for now. You can open it if you want

**Comment by rtviii** [comment]
Please put this into pydantic already. Love the project, but this has been in various stages of discussion at least since 2019 (re `@property` decorators)!

**Reply by caniko** [reply]
wait for v2.
